### PR TITLE
Refactor radio replace ripple with lazy mwc-ripple

### PR DESCRIPTION
**Pull request by bramkragten**
Refactor radio replace ripple with lazy mwc-ripple

And fix a bug where if you would focus a radio group for the second time while not focussing other radio groups it would not set focus on the selected element

### CODE REVIEW

**Comment by bramkragten** [comment]
Another option for the focus issue would be introducing a blur function:

```ts
  /**
   * Remove focused set when no element of the set has focus anymore, or if overridden by mouse
   *
   * @param element Element from which selection set is derived and subsequently
   *     blurred.
   */
  async blur(element: CheckableElement) {
    const set = this.getSet(element.name);
    const currentSelected = set.selected;
    if (!currentSelected) {
      return;
    }
    // wait for update set
    await new Promise((resolve) => setTimeout(resolve, 0));
    if (set == this.focusedSet && this.focusedSet.selected == currentSelected) {
      this.focusedSet = null;
    }
  }
```

File: packages/radio/src/single-selection-controller.ts [hunk 1]
```diff
@@ -213,15 +211,14 @@ export class SingleSelectionController {
...
    *     focused.
    */
-  focus(element: CheckableElement) {
+  async focus(element: CheckableElement) {
     // Only manage focus state when using keyboard
     if (this.mouseIsDown) {
       return;
     }
     const set = this.getSet(element.name);
-    const currentFocusedSet = this.focusedSet;
```
Comment (e111077) [inline packages/radio/src/single-selection-controller.ts:222]: Why was this removed?

Comment (bramkragten) [reply packages/radio/src/single-selection-controller.ts:222]: It is no longer needed, currently, there is a bug that if you focus an element of the set for the second time it will not move the focus to the selected item because the set that is being focused is the `focusedSet` (it is not cleared on `blur` of the set). This was probably added for when you change the value with your keyboard as the `focus` event comes before the `update` call. By awaiting a promise we check the selected value after the `update` call so we don't have to handle the value change of keyboard anymore (the focused element will be the selected element) and a second focus will also work now.

Should probably add a check to see if the element is still focused.

There are other ways to handle this, like the `blur` handler in my comment above. That will keep the `focusedSet` and will clear the `focusedSet` when you `blur` the entire set or on mouse override.

File: packages/radio/src/mwc-radio-base.ts [hunk 2]
```diff
@@ -155,6 +158,19 @@ export class RadioBase extends FormElement {
...
   }
 
+  protected rippleHandlers: RippleHandlers = new RippleHandlers(() => {
+    this.shouldRenderRipple = true;
+    return this.ripple;
+  });
+
+  protected renderRipple() {
+    if (this.shouldRenderRipple) {
+      return html`<mwc-ripple .accent="${this.checked}" .disabled="${
```
[the remark this answers is not in the document]
Comment (bramkragten) [reply packages/radio/src/mwc-radio-base.ts:168]: Got to say I hate the formatter 😄  why don't you use prettier?

File: packages/radio/src/single-selection-controller.ts (hunk 1 above)
Comment (e111077) [reply packages/radio/src/single-selection-controller.ts:222]: I can't seem to repro your issue. Is there an issue filed or can you send me a repro case?

File: packages/radio/src/mwc-radio-base.ts [hunk 3]
```diff
@@ -136,16 +142,13 @@ export class RadioBase extends FormElement {
 
   focus() {
     this.focusNative();
+    this.handleRippleFocus();
```
Comment (e111077) [inline packages/radio/src/mwc-radio-base.ts:145]: I assume this doesn't need to be added to a blur fn because blur is implicit in a radio group?

File: packages/radio/src/mwc-radio-base.ts [hunk 4]
```diff
@@ -155,6 +158,23 @@ export class RadioBase extends FormElement {
     };
   }
 
+  protected rippleHandlers: RippleHandlers = new RippleHandlers(() => {
+    this.shouldRenderRipple = true;
+    return this.ripple;
+  });
+
+  protected renderRipple() {
+    if (this.shouldRenderRipple) {
```
Comment (e111077) [inline packages/radio/src/mwc-radio-base.ts:167]: Currently this if statement is unsupported in our internal SSR tooling. Can you please refactor this fn to be the same as in checkbox as an ugly ternary? Eventually we should support this pattern, but the tooling is under development.

File: packages/radio/src/single-selection-controller.ts (hunk 1 above)
Comment (bramkragten) [reply packages/radio/src/single-selection-controller.ts:222]: You can repo it here: <https://material-components.github.io/material-components-web-components/demos/formfield/>

As you can see in the GIF, the first time the first radio gets focus it is correctly setting the focus on the selected (second) radio element, but after moving the focus back to the switch and then again to the first radio it is not set to the selected but keeps the focus on the first radio element.
radio-focus

File: packages/radio/src/mwc-radio-base.ts (hunk 3 above)
Comment (bramkragten) [reply packages/radio/src/mwc-radio-base.ts:145]: I just realized that this would not work in all cases and wondered why I didn't noticed that in my tests, I see now that it `mdc-radio` has its own focus styling `.mdc-radio .mdc-radio__background::before` & `.mdc-radio__native-control:focus + .mdc-radio__background::before`

Is this something that should be removed and replaced by the ripple or should we not use the ripple for focus?

Comment (bramkragten) [reply packages/radio/src/mwc-radio-base.ts:145]: The current styling doesn't match material, as focus is always in the accent color.
image

Comment (e111077) [reply packages/radio/src/mwc-radio-base.ts:145]: Hmm I'm going to have to pass this onto @dfreedm since he's been working with radio more-closely

Comment (dfreedm) [reply packages/radio/src/mwc-radio-base.ts:145]: I believe that styling is holdover from MDC for a CSS-only ripple. You should be able to turn it off by adding a class named `mdc-radio--upgraded` to the `mdcRoot` node

Comment (dfreedm) [reply packages/radio/src/mwc-radio-base.ts:145]: Hmm, radio works a bit differently than checkbox, the right classname is actually `mdc-ripple-upgraded`

Comment (bramkragten) [reply packages/radio/src/mwc-radio-base.ts:145]: That doesn't seem to work, but this does:
```scss
  .mdc-radio__background::before {
    content: none;
  }
```

File: packages/radio/src/single-selection-controller.ts (hunk 1 above)
Comment (bramkragten) [reply packages/radio/src/single-selection-controller.ts]: I changed it to the blur logic to not mess with setting focus, that should stay sync.

**Comment by taylor-vann** [comment]
Fixed with PR #1684
